### ISSUE TITLE
Could not get Cordova FileSystem

### ISSUE THREAD

**Issue by yeahstyle** [markmarijnissen/cordova-app-loader#52]
Hello guys!

I'm running into a little problem.

I need to use window.requestFileSystem() to get access to the local files. 
It takes four parameters:
PERSISTENT or TEMPORARY (1 or 0)
quota in bytes
success callback
fail callback.

In order to pass a quota bigger than 0mb we need to ask the user for permission.
I've read the documentation and the current non-deprecated method to do that is:
navigator.webkitPersistentStorage.requestQuota()

In an Android browser this should prompt the user requesting permission to expand the quota.
Unfortunately a Cordova Application does not run in Android Browser but a Webview and apparently- the Webview won't prompt the user for permission.

The navigator.webkitPersistentStorage.requestQuota() resolves immediately (without asking) with a granted quota of 0

The rest of the Cordova App Loader implementation fails since there's no access to the Cordova File System which returns an error code of 10, which corresponds to QUOTA_EXCEEDED_ERR

![screen shot 2015-08-05 at 21 32 48](https://cloud.githubusercontent.com/assets/1191724/9101224/8b224b52-3bba-11e5-9f4e-def65d29653a.png)

I'm using Android Emulator 5554: Nexus_5_API_21_x86

Any ideas? How do you deal with this fileSystem Quota problem?

please help!! 

thanks!

**Comment by yeahstyle** [comment]
To make this work in the emulator I had to change the quota in window.requestFileSystem() to 0.
Then I could retrieve the file system just fine! For that I had to change the following line in cordova-app-loader-complete.js:

`options.storageSize = (typeof options.storageSize !== 'undefined') ? options.storageSize : 20*1024*1024; // CHANGE: allow passing a 0`

That solved one problem.

The other problem I've been having is: 
`cdvfile://localhost/persistent/`

Whenever the root is updated in the new manifest and the files are requested using 
`cdvfile://localhost/persistent/` they can't be found. 
Cached files are being downloaded to: `file:///data/data/com.olbg/`

so I had to change the URL I use in the manifest.root. This is the line I changed in cordova-app-loader-complete.js:

`self.newManifest.root = self.cache.localUrl; // CHANGE: Use the native URL instead!`

I'm using the local URL instead of localInternalUrl.

Something tells me making these changes is wrong. (even though they work in the emulator)
Any ideas?

**Comment by dcartwright** [comment]
Your second problem maybe be an issue with URL white-listing: https://github.com/markmarijnissen/cordova-app-loader/issues/39. Also double-check that the files that are getting the 404 from cdvfile:// are actually _in_ the downloaded cache. In my case I also had a few "load" files that were not in the "files" list in the manifest, so they genuinely were not there. When you change the manifest back to the localUrl instead of localInternalUrl, I think you're basically just switching from downloaded files back to bundled files - so it _runs_, but it defeats the whole purpose of using cordova-app-loader (not loading updated files from the download).

(I don't have any insight into your problem with quotas.)

**Comment by yeahstyle** [comment]
Hello! I'm using white-list plugin and I added cdvfile://* to the Content-Security-Policy tag. 
`<meta http-equiv="Content-Security-Policy" content="default-src 'unsafe-inline' 'unsafe-eval' http://* https://* data: cdvfile://* content://* file://*;">`

The app actually doesn't find the files.

localUrl actually points to the cache folder and not the bundle one I think. The bundle folder is in:
`file:///android_asset/www/` and the new cached files are stored in `file:///data/data/com.olbg/` which is where localUrl points to.

The files are being updated (and served) correctly from server everytime I change them in the server so it is actually working. 

thanks for your insight dcartwright

**Comment by ajp8164** [comment]
yeahstyle - your OP describes precisely the issue I have run into.  Have you been able to get quota allocation?  No matter what I try I can't get the UI to present the user with an option to grant permission.  Any insight appreciated.

**Comment by yeahstyle** [comment]
Hello Andy! I managed to overcome this issue passing 0 as a value for
grantedBytes. So far my app has been working all right across several
devices.

[image: Inline image 1]

On Fri, Nov 6, 2015 at 7:19 PM, Andy Phillipson <notifications@github.com>
wrote:

> yeahstyle - your OP describes precisely the issue I have run into. Have
> you been able to get quota allocation? No matter what I try I can't get the
> UI to present the user with an option to grant permission. Any insight
> appreciated.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/markmarijnissen/cordova-app-loader/issues/52#issuecomment-154544661>
> .
>



-- 
Alvaro Casanova (Yeahstyle DG)
Web / UI Designer & Illustrator
http://www.yeahstyledg.com

**Comment by ajp8164** [comment]
Thanks for the quick response!  I've done the same, I get the file system started but when I request quota I get zero back.  A view of setup..

`      window.requestFileSystem  = window.requestFileSystem || window.webkitRequestFileSystem;

      var requestedBytes = 5*1024*1024;
      navigator.webkitPersistentStorage.requestQuota(requestedBytes, function(grantedBytes) {
        $log.debug('Granted ' + grantedBytes + ' bytes');
        window.requestFileSystem(window.PERSISTENT, grantedBytes, onFileSystemSuccess, fail);
      }, fail);`

**Comment by yeahstyle** [comment]
Passing a 0 is alright. The problem is that cordova-app-loader won't allow
passing a 0 unless you modify this line like this:

```
options.storageSize = (typeof options.storageSize !== 'undefined') ?
options.storageSize : 20*1024*1024; // CHANGE: allow passing a 0
```

I can't remember how was this line before but after I modified like this it
started to work for me. Even when passing a 0... cordova.app.loader would
end up passing  20*1024*1024.
With that conditional I wrote- now you have to option to either pass a
value or just default to storage 20*1024*1024 if you don't pass anything

On Fri, Nov 6, 2015 at 7:25 PM, Andy Phillipson <notifications@github.com>
wrote:

> Thanks for the quick response! I've done the same, I get the file system
> started but when I request quota I get zero back. A view of setup..
>
> ` window.requestFileSystem = window.requestFileSystem ||
> window.webkitRequestFileSystem;
>
>   var requestedBytes = 5*1024*1024;
>   navigator.webkitPersistentStorage.requestQuota(requestedBytes, function(grantedBytes) {
>     $log.debug('Granted ' + grantedBytes + ' bytes');
>     window.requestFileSystem(window.PERSISTENT, grantedBytes, onFileSystemSuccess, fail);
>   }, fail);`
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/markmarijnissen/cordova-app-loader/issues/52#issuecomment-154546274>
> .
>



-- 
Alvaro Casanova (Yeahstyle DG)
Web / UI Designer & Illustrator
http://www.yeahstyledg.com

**Comment by ajp8164** [comment]
I'm not using the cordova-app-loader so my solution will be different but I'm experiencing the same issue you reported; that the webview is not prompting for permission from the user.  When you made the changes did the UI start prompting for permissions from the user?

**Comment by yeahstyle** [comment]
nope! passing 0 will just start fileSystem with no problems. At least in
the devices I've tried so far. Have you tried manipulating some files
already? give it a go. you'll see if works!

On Fri, Nov 6, 2015 at 7:43 PM, Andy Phillipson <notifications@github.com>
wrote:

> I'm not using the cordova-app-loader so my solution will be different but
> I'm experiencing the same issue you reported; that the webview is not
> prompting for permission from the user. When you made the changes did the
> UI start prompting for permissions from the user?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/markmarijnissen/cordova-app-loader/issues/52#issuecomment-154550534>
> .
>



-- 
Alvaro Casanova (Yeahstyle DG)
Web / UI Designer & Illustrator
http://www.yeahstyledg.com

**Comment by mike-nelson** [comment]
Also not using cordova-app-loader but have been passing zero and it works fine up to a point. But have now had users complain they are hitting the quota error QUOTA_EXCEEDED_ERR (after using the app for a while and therefore some files have been saved to disk - not sure how many bytes). I thought zero should be unlimited. Not sure whether there is a limit and how it should prompt user to override it. Any ideas?

**Comment by domharrington** [comment]
I sporadically get this error on iOS. Wondering if a good solution would be to reject the `loader.check()` promise if `window.__fs` is undefined? At least then the errors will propagate up and the app will be able to report. At the moment when the fs is not instantiated, no errors are reported from this module.
